src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java

```diff
@@ -269,6 +277,216 @@ private Optional<Pocket> findPocketForReload(ReloadBalanceRequest reloadRequest,
+        }
+
+        OperationType previousOperationType = operationToRevert.getOperationType();
+        log.info("Operation type to revert: {}", previousOperationType);
+        if (previousOperationType.equals(RELOAD)) {
+            return revertReloadOperation(balance, revertRequest, operationToRevert);
+        } else if (previousOperationType.equals(CHARGING)) {
+            return revertChargingOperation(balance, revertRequest, operationToRevert);
+        } else {
+            log.error("Could not operation provide, because revert operation type: {} - not supported", operationToRevert.getOperationType());
+            throw new BusinessException("error.revert.operation.type.unknown", "Revert operation type not supported");
+        }
+    }
+
+    private BalanceChangeEventDto revertReloadOperation(Balance balance,
+                                                        RevertBalanceOperationRequest revertRequest,
+                                                        BalanceChangeEventDto reloadOperationToRevert) {
+        String operationUuid = StringUtils.isNotBlank(revertRequest.getUuid())
+            ? revertRequest.getUuid()
+            : UUID.randomUUID().toString();
+
+        BalanceChangeEvent changeEvent = createBalanceChangeEvent(
+            operationUuid,
+            reloadOperationToRevert.getAmountTotal(),
+            Metadata.of(revertRequest.getMetadata()),
+            now(clock),
+            balance,
+            true,
+            REVERT_RELOAD
+        );
+        changeEvent.setRevertOperationId(revertRequest.getUuidToRevert());
+
+        log.info("Revert reload mode: {}", revertRequest.getRevertReloadMode());
+        if (revertRequest.getRevertReloadMode() == null) {
+            log.error("Could not operation provide, because empty revert reload mode");
+            throw new BusinessException("error.revert.reload.mode.empty", "Empty revert reload mode");
+        } else if (revertRequest.getRevertReloadMode().equals(EXCEPTION) || revertRequest.getRevertReloadMode().equals(AS_MANY_AS_POSSIBLE_FROM_POCKETS)) {
+            revertReloadByMode(revertRequest.getRevertReloadMode(), reloadOperationToRevert.getPocketChangeEvents(), changeEvent);
+        } else if (revertRequest.getRevertReloadMode().equals(AS_MANY_AS_POSSIBLE_FROM_BALANCE)) {
+            chargingPockets(balance, reloadOperationToRevert.getAmountTotal(), changeEvent, now(clock), true, null);
+        } else {
+            log.error("Could not operation provide, because unknown revert reload mode: {}", revertRequest.getRevertReloadMode());
+            throw new BusinessException("error.revert.reload.mode.unknown", "Unknown revert reload mode");
+        }
+
+        changeEvent = balanceChangeEventRepository.save(changeEvent);
+
+        return balanceChangeEventMapper.toDto(changeEvent);
+    }
+
+    private void revertReloadByMode(RevertReloadMode revertReloadMode,
```

Comment:
if mode is as many as possible - how do you return the amount you've actually managed to revert? I don't see any place where this is calculated

Use case: 
1. User sends requests to revert reload operation, amount was 1000. 
2. System manages to revert 600. 
3. User receives response, that amountDelta is 600.

Reply:
Discussed, fix is ​​not required